### PR TITLE
Move functions which are not part of exposed C API from `mp4parse_capi` to `mp4parse` crate

### PR DESCRIPTION
**Pull request by AdrianEddy**
Move `create_sample_table`, `sample_to_chunk_iter`, `media_time_to_us`, `track_time_to_us` `Mp4parseIndice` and `Mp4parseTrackType` to `mp4parse` crate, as they are not part of the exposed C API and don't use `Mp4parseParser` C type.

This will allow users to access these functions from Rust.

There are no changes in the functions/types themselves, they are just moved around. One exception is `TrackType` type, which in C api has int values and lacks `Unknown` variant, and also the default is `Video`, while in mp4parse it's `Unknown`. What should we do with this one? And what value should Unknown have? 3? -1?

### CODE REVIEW

**Comment by AdrianEddy** [comment]
Tests are failing, because the `Default` for `TrackType` changed from `Video` to `Unknown`, I think the correct default is `Unknown`. Should I update the test in `minimal_mp4_get_track_info_invalid_track_number` to expect `Unknown` track type? This seems logical, but it changes the behavior of the C api, so just wanted to make sure.

**Comment by baumanj** [comment]
Will plan to take a look at this tomorrow. I just wanted to let you know it's still on my list

File: mp4parse/src/lib.rs
Comment (baumanj) [inline mp4parse/src/lib.rs:884]: This is a move of https://github.com/mozilla/mp4parse-rust/blob/1ab30201ad06389d0d0ebe3036418d624e7219f4/mp4parse_capi/src/lib.rs#L707-L746 with the addition of making these three functions public.

Generally, the same points as with `create_sample_table` apply, so I think we could use the same approaches, but it's less clear to me why a consumer of this crate would want to expose *these* functions. Can you talk more about your use case?

Also, I think if we're going to make time conversion functions part of a public interface, it would be preferable to (rather than using `u64` or `T: PrimInt + Zero`) either:
- Make use of a newtype like `pub struct Microseconds<T>(T);` which makes unit-mixing mistakes easier to avoid
- Return a stdlib [Duration](https://doc.rust-lang.org/stable/std/time/struct.Duration.html) value

File: mp4parse/src/lib.rs
Comment (baumanj) [inline mp4parse/src/lib.rs:916]: This is a move of https://github.com/mozilla/mp4parse-rust/blob/1ab30201ad06389d0d0ebe3036418d624e7219f4/mp4parse_capi/src/lib.rs#L2088-L2118 without any modification. I appreciate moving the tests closer to the code they cover.

File: mp4parse/src/lib.rs
Comment (baumanj) [inline mp4parse/src/lib.rs:1658]: To respond to [your comment](https://github.com/mozilla/mp4parse-rust/pull/297#issuecomment-818031114):
> Tests are failing, because the `Default` for `TrackType` changed from `Video` to `Unknown`, I think the correct default is `Unknown`. Should I update the test in `minimal_mp4_get_track_info_invalid_track_number` to expect `Unknown` track type? This seems logical, but it changes the behavior of the C api, so just wanted to make sure.

I don't think we should collapse `Mp4parseTrackType` into `TrackType`. The two are distinct, and I think that's intentional, because `TrackType` has the possibility of being the `Unknown` variant, whereas `Mp4parseTrackType` does not. Unless there's a valuable reason to expose that to the C API, I'd say it's better not to.

Do you have a use case which requires it?

**Review by baumanj — CHANGES_REQUESTED** [review]
Overall, this looks good. A number of the comments I'm making here are just to reflect my validation of what exactly is changed in the big chunks of code which have moved between files. Those comments are more about my review due diligence than anything else, so they don't need any response unless I've misstated something.

The main issue that needs to be resolved before we can merge this is the question of what public interfaces these crates will expose and what expectations consumers can have about their stability. See inline comments [[1](https://github.com/mozilla/mp4parse-rust/pull/297#discussion_r619434552)], [[2](https://github.com/mozilla/mp4parse-rust/pull/297#discussion_r619470044)], [[3](https://github.com/mozilla/mp4parse-rust/pull/297#discussion_r619473376)] for more on that, but I think we should be able to figure something out.

File: mp4parse/src/lib.rs
Comment (AdrianEddy) [reply mp4parse/src/lib.rs]: Ok, you're right, `rational_scale` doesn't need to be public. The other two are public because they are used by `mp4parse_capi` crate. My usecase doesn't involve them directly (yet), but I do see them being useful for the consumer of this crate.

As for the return type, I think stdlib `Duration` is a better choice and I can go with that, but then should we also rename the functions? `track_time_to_duration`? or just leave as is?

File: mp4parse/src/lib.rs
Comment (AdrianEddy) [reply mp4parse/src/lib.rs:680]: I added the `unstable-api` feature and also added the doc comments. Let me know if that is sufficient or I should make any changes.

BTW: Is there a cleaner way of adding `#[cfg(feature = "unstable-api")]`? Right now it's with every function, struct and import I moved and it doesn't look very clean in my opinion.

**Comment by baumanj** [comment]
Thanks for the update! I should be able to take a look at this tomorrow or Wednesday, hopefully

**Comment by baumanj** [comment]
Today is looking less likely, but hopefully by the end of the week.

**Comment by baumanj** [comment]
I'm looking at this now and will hopefully have some additional feedback for you today, but in the meantime, I enabled the CI run and wanted to point out this one clippy failure:
```
error: this creates an owned instance just for comparison
    --> mp4parse/src/lib.rs:2593:25
     |
2593 |                 assert!(<$inner>::from(value) <= Self::MAX);
     |                         ^^^^^^^^^^^^^^^^^^^^^ help: try: `value`
...
2616 |     impl_bounded_product!(U32MulU8, U32, U8, u64);
     |     ---------------------------------------------- in this macro invocation
     |
     = note: `-D clippy::cmp-owned` implied by `-D warnings`
     = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cmp_owned
     = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

Isn't your issue, so don't worry about it.

File: mp4parse/src/lib.rs
Comment (baumanj) [reply mp4parse/src/lib.rs]: > As for the return type, I think stdlib `Duration` is a better choice and I can go with that, but then should we also rename the functions? `track_time_to_duration`? or just leave as is?

If it's not necessary to expose `media_time_to_us` and `track_time_to_us` for your use case, I'll leave it up to you to choose:
1. Keep them private
2. Change the return on the public (still behind `unstable-api` feature) functions to `Duration` (renaming them `track_time_to_duration` and `media_time_to_duration`), and leave `mp4parse_capi::{media_time_to_us, track_time_to_us}` as private functions, implemented in terms of the new `Duration`-returning ones.

File: mp4parse/src/lib.rs
Comment (baumanj) [reply mp4parse/src/lib.rs:680]: > BTW: Is there a cleaner way of adding `#[cfg(feature = "unstable-api")]`? Right now it's with every function, struct and import I moved and it doesn't look very clean in my opinion.

I'd suggest moving all that code to a new `unstable.rs` with
```
#![cfg(feature = "unstable-api")]
```
at the top. Then add
```
#[cfg(feature = "unstable-api")]
mod unstable;
```
to lib.rs. I think that should cut way down on how many places need `#[cfg(feature = "unstable-api")]`.

**Comment by baumanj** [comment]
I think I responded to all your questions. Let me know if there's anything else you need, or else I'll take a look when you have another update.

File: mp4parse/src/lib.rs
Comment (AdrianEddy) [reply mp4parse/src/lib.rs]: Ok I went with newtype `Microseconds<T>` because stdlib `Duration` uses `u128` to represent micros and I didn't want to add that overhead

File: mp4parse/src/lib.rs
Comment (AdrianEddy) [reply mp4parse/src/lib.rs]: Moved to `unstable` mod, much cleaner now 👍

**Comment by baumanj** [comment]
> I made all requested changes, please check if everything is good now

I'm in the process of giving this a last look over, but I have some things that come first, so it may be a few days yet. Just wanted to let you know I haven't forgotten

**Review by baumanj — APPROVED** [review]
This is plenty good enough to accept. There are a few tiny tweaks that I'd like your input on before merging, but if I don't hear back I'll likely merge this as is and possibly make those changes after the fact.

In the meantime, I'll just need to confirm this doesn't introduce any issue upstream in gecko that are discoverable with CI and then I'll merge.

File: mp4parse/src/unstable.rs
Comment (AdrianEddy) [reply mp4parse/src/unstable.rs:493]: Debug and PartialEq are needed for tests, otherwise this fails to compile: 
```
assert_eq!(
    track_time_to_us(duration, scale),
    Some(Microseconds(100_079_991_719_000_000u64))
);
```

File: mp4parse_capi/src/lib.rs
Comment (AdrianEddy) [reply mp4parse_capi/src/lib.rs:624]: No reason, it was an leftover from my testing, I changed it back now 👍

**Comment by AdrianEddy** [comment]
Thanks for the review, I updated the code with all requested changes so it should be good to merge 🚀

**Comment by baumanj** [comment]
Thanks for the contribution, the excellent work and your patience!

**Pull request merged**
